### PR TITLE
Hamiltonian Monte Carlo with Energy Conserving Subsampling

### PR DESCRIPTION
**Pull request by OlaRonning**
Introduces [HMCECS](https://jmlr.csail.mit.edu/papers/volume20/17-452/17-452.pdf) to numpyro.

This PR includes:
 - Taylor proxy [section 4.2];
 - Difference estimation method [section 4.1].

### CODE REVIEW

File: examples/covtype.py [hunk 1]
```diff
@@ -30,25 +34,74 @@ def _load_dataset():
     print("Data shape:", features.shape)
     print("Label distribution: {} has label 1, {} has label 0"
           .format(labels.sum(), N - labels.sum()))
-    return features, labels
+    return features[::5], labels[::5]
```
Comment (fehiepsi) [inline examples/covtype.py:37]: I think we can use the whole dataset for the current algorithm.

**Review by fehiepsi — COMMENTED** [review]
Awesome work, @OlaRonning and @LysSanzMoreta!

There are just a few nits below. I'll take another pass to not have to trace the model several time, hence we can [optimize the compiling time](https://github.com/pyro-ppl/numpyro/issues/585). Do you want to address some TODOs in this PR or defer it to another PR?

**Comment by OlaRonning** [comment]
@fehiepsi thanks for the review. I'll address the TODOs in this PR. 

Can you come with some suggestions for test cases? Still not that comfortable with unit testing stochastic functions.

**Comment by fehiepsi** [comment]
> Can you come with some suggestions for test cases?

I think we can test if `taylor_proxy` returns correct results and `estimate_likelihood` returns values that are close to the true ~log~likelihood with taylor_proxy (in the sense that variance is small if we change subsample indices and latent values). What do you think?

File: examples/covtype.py (hunk 1 above)
Comment (OlaRonning) [reply examples/covtype.py:37]: Agree.

File: numpyro/distributions/continuous.py [hunk 2]
```diff
@@ -962,7 +962,7 @@ def sample(self, key, sample_shape=()):
 
     @validate_sample
     def log_prob(self, value):
-        normalize_term = jnp.log(jnp.sqrt(2 * jnp.pi) * self.scale)
+        normalize_term = jnp.log(jnp.sqrt(2 * jnp.pi) * self.scale)  # TODO:Added jnp.abs
```
[the remark this answers is not in the document]
Comment (OlaRonning) [reply numpyro/distributions/continuous.py:965]: Don't think this is necessary as the [constraint on `self.scale`](https://github.com/aleatory-science/numpyro/blob/feature/ecs/numpyro/distributions/continuous.py#L949) is already positive. @LysSanzMoreta let me know if you disagree.

File: numpyro/handlers.py [hunk 3]
```diff
@@ -76,13 +75,12 @@
    -874.89813
 """
 
-from collections import OrderedDict
 import warnings
+from collections import OrderedDict
```
[the remark this answers is not in the document]
Comment (OlaRonning) [reply numpyro/handlers.py:79]: Yes, of course.

File: numpyro/primitives.py [hunk 4]
```diff
@@ -373,7 +374,7 @@ def postprocess_message(self, msg):
                         raise ValueError(
                             "Inside numpyro.plate({}, {}, dim={}) invalid shape of {}: {}"
                             .format(self.name, self.size, self.dim, statement, shape))
-                    if self.subsample_size < self.size:
+                    elif self.subsample_size < self.size:
```
[the remark this answers is not in the document]
Comment (OlaRonning) [reply numpyro/primitives.py:377]: Updated, though you are right neither would enter `elif self.subsample_size < self.size` branch if `shape[dim] != self.size`.

File: numpyro/infer/hmc_gibbs.py [hunk 5]
```diff
@@ -397,31 +399,65 @@ def potential_fn(z_gibbs, z_hmc):
...
+        new_idxs[name] = new_idx
+        starts[name] = start
+    return u_new, pads, new_idxs, starts
 
-    return gibbs_fn
+
+HMCECSState = namedtuple("HMCECSState", "z, hmc_state, rng_key, gibbs_state, accept_prob")
+TaylorProxyState = namedtuple("TaylorProxyState", "ref_subsample_log_liks, "
+                                                  "ref_subsample_log_lik_grads, ref_subsample_log_lik_hessians")
+BlockPoissonEstState = namedtuple("BlockPoissonEstState", "block_rng_keys, sign")
```
[the remark this answers is not in the document]
Comment (OlaRonning) [reply numpyro/infer/hmc_gibbs.py:450]: I'll take it out. I will suggest implementing the Block Poisson Est as a Bachelor project for next semester if you've not finished the implementation.

File: numpyro/infer/hmc_gibbs.py [hunk 6]
```diff
@@ -476,48 +512,292 @@ class HMCECS(HMCGibbs):
         >>> assert abs(jnp.mean(samples) - 1.) < 0.1
 
     """
-    def __init__(self, inner_kernel, *, num_blocks=1):
+
+    def __init__(self, inner_kernel, *, num_blocks=1, proxy=None, method='perturbed'):
```
[the remark this answers is not in the document]
Comment (OlaRonning) [reply numpyro/infer/hmc_gibbs.py:516]: That makes sense!

**Comment by OlaRonning** [comment]
@fehiepsi those tests sound good. I'll add them in.

**Comment by OlaRonning** [comment]
@fehiepsi should the python version in [CI](https://github.com/pyro-ppl/numpyro/blob/master/.github/workflows/ci.yml#L17) be bumped to 3.9? looks like it's supported by [Github Actions](https://github.com/actions/setup-python/issues/148).

File: numpyro/infer/hmc_gibbs.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/infer/hmc_gibbs.py]: > implementing the Block Poisson Est as a Bachelor project for next semester

That would be great! I think I will be lazy for that task. =.=

**Comment by fehiepsi** [comment]
> should the python version in CI be bumped to 3.9

It is not needed. We typically want to test for the oldest Python version that we support.

File: numpyro/primitives.py [hunk 7]
```diff
@@ -304,10 +304,11 @@ def _subsample(name, size, subsample_size, dim):
         }
         apply_stack(msg)
         subsample = msg['value']
+        subsample_size = msg['args'][1]  # TODO: rewrite plate
```
Comment (fehiepsi) [inline numpyro/primitives.py:307]: I guess this is not relevant to this PR but this change could be useful and seems to make plate message more flexible. +1

Btw, what does this TODO mean?

File: numpyro/infer/hmc_gibbs.py [hunk 8]
```diff
@@ -476,48 +499,289 @@ class HMCECS(HMCGibbs):
         >>> assert abs(jnp.mean(samples) - 1.) < 0.1
 
     """
-    def __init__(self, inner_kernel, *, num_blocks=1):
+
+    def __init__(self, inner_kernel, *, num_blocks=1, proxy=None):
```
Comment (fehiepsi) [inline numpyro/infer/hmc_gibbs.py:503]: Could you add docs for `proxy`?

File: numpyro/examples/datasets.py [hunk 9]
```diff
@@ -240,6 +237,21 @@ def _load_jsb_chorales():
...
+
+    file_path = os.path.join(DATA_DIR, 'HIGGS.csv.gz')
+    with io.TextIOWrapper(gzip.open(file_path, 'rb')) as f:
+        csv_reader = csv.reader(f, delimiter=',', quoting=csv.QUOTE_NONE)
+        obs = []
+        data = []
+        for row in csv_reader:
+            obs.append(row[0])
+            data.append(row[1:])
+    return np.stack(obs), np.stack(data)
```
Comment (fehiepsi) [inline numpyro/examples/datasets.py:252]: Do you intend to add an example? :D

File: numpyro/infer/hmc_gibbs.py [hunk 10]
```diff
@@ -476,48 +499,289 @@ class HMCECS(HMCGibbs):
...
+        model_kwargs["_gibbs_sites"] = z_gibbs
+        model_kwargs["_gibbs_state"] = gibbs_state
+        hmc_state = self.inner_kernel.sample(hmc_state, model_args, model_kwargs)
 
         z = {**z_gibbs, **hmc_state.z}
+        return HMCECSState(z, hmc_state, rng_key, gibbs_state, accept_prob)
 
-        return HMCGibbsState(z, hmc_state, rng_key)
+
+def perturbed_method(rng_key, model, model_args, model_kwargs, proxy_fn):
```
[the remark this answers is not in the document]
Comment (OlaRonning) [reply numpyro/infer/hmc_gibbs.py:587]: Yes, makes perfect sense. I'll refactor it.

File: numpyro/infer/hmc_gibbs.py [hunk 11]
```diff
@@ -476,48 +499,289 @@ class HMCECS(HMCGibbs):
...
+        assert method is not None
+        super().__init__(fn)
+        self.method = method
+        self.params = None
+        self.likelihoods = {}
+        self.subsample_plates = {}
+        self.gibbs_state = None
+
+    def __enter__(self):
+        # trace(substitute(substitute(control_variate(model), unconstrained_reparam)))
```
[the remark this answers is not in the document]
Comment (OlaRonning) [reply numpyro/infer/hmc_gibbs.py:739]: thanks, I'll update this.

File: numpyro/primitives.py (hunk 7 above)
Comment (OlaRonning) [reply numpyro/primitives.py:307]: I originally added it to use the `change_subsample_size` handler to avoid raising an error (current fix is to change to [warning](https://github.com/pyro-ppl/numpyro/blob/2c261739513eae2205e7ae17ccfa8ffeca991371/numpyro/primitives.py#L309)).

The TODO has been resolved and not relevant to current implementation. I'll remove it.

File: numpyro/examples/datasets.py (hunk 9 above)
Comment (OlaRonning) [reply numpyro/examples/datasets.py:252]: Yes, I will add a logistic regression for the HIGGS dataset with HMCECS. Maybe I'll add Consensus as well. Depends on other obligations.

File: numpyro/infer/hmc_gibbs.py [hunk 12]
```diff
@@ -476,48 +503,290 @@ class HMCECS(HMCGibbs):
...
+                      if name in reference_params}
+        ref_params_flat, unravel_fn = ravel_pytree(ref_params)
+
+        def log_likelihood(params_flat, subsample_indices=None):
+            if subsample_indices is None:
+                subsample_indices = {k: jnp.arange(v[0]) for k, v in subsample_plate_sizes.items()}
+            params = unravel_fn(params_flat)
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+                with block(), trace() as tr, substitute(data=subsample_indices), substitute(data=params):
```
Comment (fehiepsi) [inline numpyro/infer/hmc_gibbs.py:643]: Do I understand correctly that `params_flat` is unconstrained params, which are params that we want to sample from MCMC? If so, we should convert params to constrained values. One way is to use `_unconstrained_reparams`, but you can also use
```
params = {name: biject_to(prototype_trace[name]["fn"].support)(value)
          for name, value in params.items()}
```
to save some unnecessary `log_abs_det_jacobian` terms in `_unconstrained_reparam`.

File: numpyro/infer/hmc_gibbs.py [hunk 13]
```diff
@@ -476,48 +503,290 @@ class HMCECS(HMCGibbs):
...
+            log_lik_sum += unbiased_log_lik - 0.5 * variance
+        return log_lik_sum
+
+    return estimator
+
+
+def taylor_proxy(reference_params):
+    def construct_proxy_fn(prototype_trace, subsample_plate_sizes, model, model_args, model_kwargs, num_blocks=1):
+        ref_params = {name: _unconstrain_reparam(reference_params, site) for name, site in prototype_trace.items()
+                      if name in reference_params}
```
[the remark this answers is not in the document]
Comment (OlaRonning) [reply numpyro/infer/hmc_gibbs.py:634]: I see, must have miss read the docs. I'll update it.

File: setup.py [hunk 14]
```diff
@@ -35,7 +35,7 @@
         # TODO: pin to a specific version for the release (until JAX's API becomes stable)
         'jax==0.2.8',
         # check min version here: https://github.com/google/jax/blob/master/jax/lib/__init__.py#L26
-        'jaxlib==0.1.59',
+        'jaxlib==0.1.60',
```
[the remark this answers is not in the document]
Comment (OlaRonning) [reply setup.py:38]: sure, version 0.1.59 produces some errors in the subsample handler, which is why I bumped it.

File: numpyro/examples/datasets.py [hunk 15]
```diff
@@ -86,7 +83,7 @@ def _load_baseball():
     def train_test_split(file):
         train, test, player_names = [], [], []
         with open(file, 'r') as f:
-            csv_reader = csv.DictReader(f, delimiter='\t', quoting=csv.QUOTE_NONE)
+            csv_reader = csv.reader(f, delimiter='\t', quoting=csv.QUOTE_NONE)
```
[the remark this answers is not in the document]
Comment (OlaRonning) [reply numpyro/examples/datasets.py:86]: This should not have been changed I'll revert it.

[the remark this answers is not in the document]
Comment (OlaRonning) [reply numpyro/examples/datasets.py:86]: This should not have been changed I'll revert it.

File: numpyro/infer/hmc_gibbs.py (hunk 12 above)
Comment (OlaRonning) [reply numpyro/infer/hmc_gibbs.py:643]: Make sense, I'll revert it.

File: numpyro/infer/hmc_gibbs.py [hunk 16]
```diff
@@ -448,9 +472,13 @@ class HMCECS(HMCGibbs):
        Quiroz, M., Kohn, R., Villani, M., & Tran, M. N. (2018)
     3. *The Block Pseudo-Margional Sampler*,
        Tran, M.-N., Kohn, R., Quiroz, M. Villani, M. (2017)
+    4. *The Fundamental Incompatibility ofScalable Hamiltonian Monte Carlo and Naive Data Subsampling*
```
Comment (fehiepsi) [inline numpyro/infer/hmc_gibbs.py:475]: nit: of Scalable

File: numpyro/infer/hmc_gibbs.py [hunk 17]
```diff
@@ -476,48 +504,297 @@ class HMCECS(HMCGibbs):
...
+
+        def log_likelihood(params_flat, subsample_indices=None):
+            if subsample_indices is None:
+                subsample_indices = {k: jnp.arange(v[0]) for k, v in subsample_plate_sizes.items()}
+            params = unravel_fn(params_flat)
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+
+                with block(), trace() as tr, substitute(data=subsample_indices), substitute(data=params), \
+                        substitute(substitute_fn=partial(_unconstrain_reparam, params)):
```
Comment (fehiepsi) [inline numpyro/infer/hmc_gibbs.py:651]: can we use
```
params = {name: biject_to(prototype_trace[name]["fn"].support)(value) for name, value in params.items()}
```
here so you can keep `substitute(data=params)` and remove this `substitute_fn`. I don't know how it works when we use both `data=params` and `substitute_fn=...` here.

File: numpyro/infer/hmc_gibbs.py [hunk 18]
```diff
@@ -476,48 +504,297 @@ class HMCECS(HMCGibbs):
...
     """
-    def __init__(self, inner_kernel, *, num_blocks=1):
+
+    def __init__(self, inner_kernel, *, num_blocks=1, proxy=None):
         super().__init__(inner_kernel, lambda *args: None, None)
+
+        self.inner_kernel._model = _wrap_gibbs_state(self.inner_kernel._model)
         self._num_blocks = num_blocks
+        self._proxy = proxy
+        self._grad_ = None
```
Comment (fehiepsi) [inline numpyro/infer/hmc_gibbs.py:514]: nit: this is not needed

File: numpyro/infer/hmc_gibbs.py [hunk 19]
```diff
@@ -476,48 +504,297 @@ class HMCECS(HMCGibbs):
...
+        model_kwargs["_gibbs_sites"] = z_gibbs
+        model_kwargs["_gibbs_state"] = gibbs_state
+        hmc_state = self.inner_kernel.sample(hmc_state, model_args, model_kwargs)
 
         z = {**z_gibbs, **hmc_state.z}
+        return HMCECSState(z, hmc_state, rng_key, gibbs_state, accept_prob)
 
-        return HMCGibbsState(z, hmc_state, rng_key)
+    @property
+    def grad_(self):
```
Comment (fehiepsi) [inline numpyro/infer/hmc_gibbs.py:592]: nit: this property is not needed

**Review by fehiepsi — COMMENTED** [review]
Looks clean to me! Thanks for addressing my comments, @OlaRonning! Please let me know if you want to add more tests in another PR or if this is ready to merge when the following comments are addressed.

File: numpyro/distributions/continuous.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/distributions/continuous.py:965]: Please remove this TODO too. I think the math here is right.

**Comment by OlaRonning** [comment]
Hi @fehiepsi, thanks for the reviews! 

I'll add the test cases (proxy result and low variance) you suggested and then it's ready to be merged.

File: numpyro/infer/hmc_gibbs.py (hunk 17 above)
Comment (OlaRonning) [reply numpyro/infer/hmc_gibbs.py]: I'll update this.

**Review by fehiepsi — APPROVED** [review]
